libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Conway/ParametricSpec.hs

```diff
@@ -0,0 +1,301 @@
+                (branch $ \_ -> True)
+        )
+    ]
+
+-- | Generate random Stake references that have a high probability of being delegated.
+delegatedStakeReference ::
+  (IsConwayUniv fn, Crypto c) =>
+  Term fn (Map (Credential 'Staking c) (KeyHash 'StakePool c)) ->
+  Specification fn (StakeReference c)
+delegatedStakeReference delegs =
+  constrained $ \ [var|ref|] ->
+    caseOn
+      ref
+      (branchW 9 $ \ [var|base|] -> member_ base (dom_ delegs))
+      (branchW 0 $ \_ptr -> False)
+      (branchW 1 $ \_null -> True) -- just an occaisional NullRef
+
+instantaneousRewardsSpec ::
+  forall c fn.
+  (IsConwayUniv fn, Crypto c) =>
+  Term fn AccountState ->
+  Specification fn (InstantaneousRewards c)
+instantaneousRewardsSpec acct = constrained $ \ [var| irewards |] ->
+  match acct $ \ [var| acctRes |] [var| acctTreas |] ->
+    match irewards $ \ [var| reserves |] [var| treasury |] [var| deltaRes |] [var| deltaTreas |] ->
+      [ dependsOn acctRes reserves
+      , dependsOn acctRes deltaRes
+      , dependsOn acctTreas treasury
+      , dependsOn acctTreas deltaTreas
+      , assertExplain (pure "deltaTreausry and deltaReserves sum to 0") $ negate deltaRes ==. deltaTreas
+      , forAll (rng_ reserves) (\ [var| x |] -> x >=. (lit (Coin 0)))
+      , forAll (rng_ treasury) (\ [var| y |] -> y >=. (lit (Coin 0)))
+      , assert $ (toDelta_ (foldMap_ id (rng_ reserves))) - deltaRes <=. toDelta_ acctRes
+      , assert $ (toDelta_ (foldMap_ id (rng_ treasury))) - deltaTreas <=. toDelta_ acctTreas
+      ]
+
+instance IsConwayUniv fn => EraOut Shelley fn where
+  irewardSpec = instantaneousRewardsSpec
+  hasPtrs _proxy = lit True
+  correctTxOut = betterTxOutShelley
+
+instance IsConwayUniv fn => EraOut Allegra fn where
+  irewardSpec = instantaneousRewardsSpec
+  hasPtrs _proxy = lit True
+  correctTxOut = betterTxOutShelley
+
+instance IsConwayUniv fn => EraOut Mary fn where
+  irewardSpec = instantaneousRewardsSpec
+  hasPtrs _proxy = lit True
+  correctTxOut = betterTxOutMary
+
+instance IsConwayUniv fn => EraOut Alonzo fn where
+  irewardSpec = instantaneousRewardsSpec
+  hasPtrs _proxy = lit True
+  correctTxOut = betterTxOutAlonzo
+
+instance IsConwayUniv fn => EraOut Babbage fn where
+  irewardSpec = instantaneousRewardsSpec
+  hasPtrs _proxy = lit True
+  correctTxOut = betterTxOutBabbage
+
+instance IsConwayUniv fn => EraOut Conway fn where
+  irewardSpec _ = constrained $ \ [var|irewards|] ->
+    match irewards $ \ [var|reserves|] [var|treasury|] [var|deltaRes|] [var|deltaTreas|] ->
+      [ reserves ==. lit Map.empty
+      , treasury ==. lit Map.empty
+      , deltaRes ==. lit (DeltaCoin 0)
+      , deltaTreas ==. lit (DeltaCoin 0)
+      ]
+  hasPtrs _proxy = lit False
+  correctTxOut = betterTxOutBabbage
+
+-- =========================================================================
+-- Making Cert Era parametric with the EraCert class
+
+class (Era era, HasSpec ConwayFn (TxCert era)) => EraCert era where
```

Comment:
I think this should also have a better name

Reply:
done